### PR TITLE
Support mapping the Ballerina external function args to jBallerina BArray

### LINKED ISSUE #42404: [Improvement]: Map the Ballerina external function arguments to Java varargs
### Description

This is a requirement for this proposal: https://github.com/ballerina-platform/ballerina-library/issues/6099

There is a requirement to access a Ballerina external function arguments as Java varargs. This is because the external functions are generated by the OpenAPI tool and it can have arbitrary function parameters.

### Describe your problem(s)

For example see the below generated code:
```bal
public isolated client class Client {

   ...

    resource isolated function get albums/[int id](typedesc<Album|AlbumFound|AlbumNotFound> targetType = <>) 
            returns targetType|error = @java:Method {
        'class: "io.ballerina.openapi.client.GeneratedClient",
        name: "getResource"
    } external;

    resource isolated function get albums(string genre, typedesc<Album[]|AlbumsFound|NoAlbumFound> targetType = <>) 
            returns targetType|error = @java:Method {
        'class: "io.ballerina.openapi.client.GeneratedClient",
        name: "getResource"
    } external;

    ...
}
```

Note that both external functions call the same Java function. This is by design, so that the Java code is static and will be as a Java dependency at build time.

Since the generated code can have arbitrary arguments, the Java function signature is defined like this:
```java
public static Object getResource(Environment env, BObject client, BArray path, Object... values) {
   ....
}
```

> **Note:** The env and client is auto injected. The `BArray path` represents the path parameters of the resource function.

When invoked with the above Java code, the following error is thrown at compile time:
```console
ERROR [client.bal:(41:5,45:16)] {ballerina/jballerina.java}METHOD_SIGNATURE_DOES_NOT_MATCH 'Incompatible param type for method 'getResource' in class 'io.ballerina.openapi.client.GeneratedClient': Java type '[Ljava.lang.Object;' will not be matched to ballerina type 'typedesc<(tharmigan/status_code_binding:0.1.0:Album|tharmigan/status_code_binding:0.1.0:AlbumFound|tharmigan/status_code_binding:0.1.0:AlbumNotFound)>''
```

### Describe your solution(s)

Add support for mapping the Ballerina external function arguments to Java varargs.

### Related area

-> Compilation

### Related issue(s) (optional)

- https://github.com/ballerina-platform/ballerina-library/issues/6099

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

@warunalakshitha @gabilang

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #42404

## Approach
Bundled extern function args into `any[]` and mapped to jBallerina `BArray`. It's currently support for resource and remote methods. 

## Samples
```ballerina
public isolated client class ClientObj {
    isolated resource function get orderitem/[string orderId]/[string itemId](typedesc<anydata> targetType = <>)
                                                            returns targetType|error = @java:Method {
        'class: "org/ballerinalang/nativeimpl/jvm/tests/StaticMethods",
        name: "getResource",
        paramTypes: ["io.ballerina.runtime.api.Environment", "io.ballerina.runtime.api.values.BObject",
                                    "io.ballerina.runtime.api.values.BArray", "io.ballerina.runtime.api.values.BArray"]
    } external;

    isolated resource function get orderitem/[string id](int i, float f, decimal d, string s,
                                        typedesc<anydata> targetType = <>) returns targetType|error = @java:Method {
        'class: "org/ballerinalang/nativeimpl/jvm/tests/StaticMethods",
        name: "getResource",
        paramTypes: ["io.ballerina.runtime.api.Environment", "io.ballerina.runtime.api.values.BObject",
                                    "io.ballerina.runtime.api.values.BArray", "io.ballerina.runtime.api.values.BArray"]
    } external;
}
```
Above resource functions can be mapped to a single Java method 
```java
public static Object getResource(Environment env, BObject client, BArray path, BArray args) {
    return 1;
}
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by warunalakshitha** [comment]
Shall we improve the coverage as well. Lets add some negative cases as well

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java [hunk 1]
```diff
@@ -183,25 +187,97 @@ private boolean hasEqualParamCounts(JMethodRequest jMethodRequest, JMethod jMeth
...
+                jMethod.setReceiverType(jMethodRequest.receiverType);
+            }
+            return jMethod.isBalEnvAcceptingMethod();
+        }
+        return false;
+    }
+
+    private boolean isAcceptingBundledFunctionParamsOnly(JMethodRequest jMethodRequest, JMethod jMethod,
+                                                         int reducedParamCount) {
+        int count = jMethod.getParamTypes().length;
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:234]: Shall we add some code comments in this logic

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java [hunk 2]
```diff
@@ -183,25 +187,97 @@ private boolean hasEqualParamCounts(JMethodRequest jMethodRequest, JMethod jMeth
...
+                jMethod.setReceiverType(jMethodRequest.receiverType);
+            }
+            return jMethod.isBalEnvAcceptingMethod();
+        }
+        return false;
+    }
+
+    private boolean isAcceptingBundledParameters(JMethodRequest jMethodRequest, JMethod jMethod) {
+        int count = jMethod.getParamTypes().length;
+        if (count < 2 || count > 4 || jMethodRequest.pathParamCount == 0) {
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:265]: Lets add some comments why we check 2 and 4

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java [hunk 3]
```diff
@@ -963,6 +1068,37 @@ private boolean isFirstPathParamARestParam(JMethodRequest jMethodRequest, JMetho
...
+    }
+
+    private boolean checkAllParamsAnydataAssignable(JMethodRequest jMethodRequest) {
+        if (jMethodRequest.kind != JMethodKind.METHOD) {
+            return false;
+        }
+        for (int i = 0; i < jMethodRequest.bParamTypes.length; i++) {
+            BType bType = jMethodRequest.bParamTypes[i];
+            switch (Types.getImpliedType(bType).tag) {
+                case TypeTags.INT, TypeTags.BYTE, TypeTags.FLOAT, TypeTags.DECIMAL, TypeTags.STRING,
```
[the remark this answers is not in the document]
Comment (gabilang) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java:1086]: It uses `compiler.utils.TypeTags` and it does not pertain an order to give it as a tag range.

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**